src/main/java/org/oscii/morph/Substitutor.java

```diff
@@ -0,0 +1,170 @@
+        if (vocab == null) {
+            vocab = corpus.vocab();
+        }
+        Stream<RuleLexicalized> p = IndexByStem(vocab, Substitutor::getPrefix).values()
+                .parallelStream().flatMap(x -> transformations(x, Rule.Prefix::new));
+        Stream<RuleLexicalized> s = IndexByStem(vocab, Substitutor::getSuffix).values()
+                .parallelStream().flatMap(x -> transformations(x, Rule.Suffix::new));
+        substitutions = Stream.concat(p, s).collect(groupingBy(t -> t.sub));
+    }
+
+    /**
+     * Limit the splits for each word pair to the K most frequent substitutions and keeps only
+     * substitutions supported by a minimum number of word pairs.
+     * <p>
+     * A word pair is split by a substitution rule. Multiple rules can split the
+     * same word pair. E.g., "wind" and "winded" can be split by
+     * <p>
+     * s/ε/ed, s/d/ded, s/nd/nded, s/ind/inded, and p/win/winde
+     * <p>
+     * Pruning to the most frequent substitution would keep e.g. only s/ε/ed.
+     * <p>
+     * This pruning step is not part of Soricut & Och 2015.
+     *
+     * @param maxSplitsPerPair k most frequent splits to keep
+     */
+    public void prune(int maxSplitsPerPair, int minPairCount) {
+        Map<Rule, List<RuleLexicalized>> a, b, c;
+
+        log.info("Pruning substitution rules: minPairCount of {}", minPairCount);
+        a = enforceMinPairCount(substitutions, minPairCount);
+
+        log.info("Pruning substitution rules: maxSplitsPerPair of {}", maxSplitsPerPair);
+        b = a.values()
+                .parallelStream().flatMap(ts -> ts.stream())
+                .collect(groupingBy(t -> t.pair))
+                .values().parallelStream()
+                .flatMap(ts -> mostFrequent(ts, maxSplitsPerPair))
+                .collect(groupingBy(t -> t.sub));
+
+        log.info("Pruning substitution rules: minPairCount of {} (again)", minPairCount);
+        c = enforceMinPairCount(b, minPairCount);
+
+        substitutions = c;
+    }
+
+    public void scoreRules(RuleScored.ScoringParams params) {
+        log.info("Scoring substitutions");
+        scored = substitutions.entrySet().parallelStream()
+                .map(e -> scoreRule(e.getKey(), e.getValue(), params))
+                .sorted((r, s) -> Double.compare(s.hitRate, r.hitRate))
+                .collect(toList());
+    }
+
+    private RuleScored scoreRule(Rule rule, List<RuleLexicalized> support, RuleScored.ScoringParams params) {
+        RuleScored rs = new RuleScored(rule, support);
+        rs.score(embeddings, params);
+        return rs;
+    }
+
+    private Stream<RuleLexicalized> mostFrequent(List<RuleLexicalized> ts, int k) {
+        if (ts.size() <= k) return ts.stream();
+        return ts.stream().sorted(comparingInt(t -> substitutions.get(t.sub).size())).limit(k);
+    }
+
+    // All transformations for a list of segmentations with a common stem.
+    private Stream<RuleLexicalized> transformations(List<Segmentation> segs,
```

Comment:
I'd prefer renaming this method, e.g. "getRulesLexicalized"; "transformations" suggests Stream<Transformation> as the return type instead of Stream<RuleLexicalized>